### PR TITLE
make it possible to ifdef out statistics

### PR DESCRIPTION
**Pull request by ldorau**
This change is [Reviewable](https://reviewable.io/reviews/pmem/vmemcache/163)

### CODE REVIEW

**Comment by kilobyte** [comment]
I wonder, perhaps a runtime flag would be fast enough?  That'd cost us an `if` and a read from a cacheline that's already hot — negligible compared to other overheads we have.

A full-bus-lock instruction is nasty and worth avoiding, yeah.

**Comment by ldorau** [comment]
> I wonder, perhaps a runtime flag would be fast enough? That'd cost us an `if` and a read from a cacheline that's already hot — negligible compared to other overheads we have.
> 

But if someone turns statistics off and on again, they will be incorrect.

File: doc/vmemcache.md [hunk 1]
```diff
@@ -204,6 +204,8 @@ Obtains a piece of statistics about the cache. The *stat* may be:
  + **VMEMCACHE_STAT_POOL_SIZE_USED**
 	current usage of data pool
 
+Statistics are enabled by default. They can be disabled at the compile time
```
Comment (pbalcer) [inline doc/vmemcache.md]: This should be in README, not man page.

File: src/vmemcache.c [hunk 2]
```diff
@@ -277,7 +277,9 @@ vmemcache_put(VMEMcache *cache, const void *key, size_t ksize,
 		return -1;
 	}
 
+#ifndef VMEMCACHE_NO_STATS
 	util_fetch_and_add64(&cache->size_DRAM, malloc_usable_size(entry));
```
Comment (pbalcer) [inline src/vmemcache.c]: just make `stat_get`, `stat_set` wrappers which have two implementations.
https://github.com/pmem/pmdk/blob/master/src/libpmemobj/stats.h

File: doc/vmemcache.md (hunk 1 above)
Comment (ldorau) [reply doc/vmemcache.md]: I think it should be in both of them. Added to README.

File: src/vmemcache.c (hunk 2 above)
Comment (ldorau) [reply src/vmemcache.c]: Done

**Comment by ldorau** [comment]
@kilobyte please review

**Comment by kilobyte** [comment]
The code looks sound, but I'm afraid that the vast majority of users, who use distro compiled binaries, will get the locking penalty in all cases.

Most of the data is already behind a lock; updating a variable there is near-free (the processor can reorder an update that doesn't affect other work).  The only stats not behind an exclusive lock are gets/hits/misses; not sure how to update those.

But it's probably better to merge your code now and only then optimize one-by-one.

**Comment by ldorau** [comment]
@kilobyte @pbalcer please resolve your discussions

File: benchmarks/bench_simul.c [hunk 3]
```diff
@@ -157,6 +157,13 @@ static struct param_t {
 	{ 0 },
 };
 
+static struct {
+	os_cond_t cond;
+	os_mutex_t mutex;
+	uint64_t wanted;
+} ready;
+
+#ifndef VMEMCACHE_NO_STATS
```
Comment (marcinslusarz) [inline benchmarks/bench_simul.c]: negation of negation...

Please reverse the logic.

File: README.md [hunk 4]
```diff
@@ -30,3 +30,10 @@ $ cmake .. -DCMAKE_INSTALL_PREFIX=~/vmemcache-bin
 $ make
 $ make install
 ```
+
+# Statistics #
+
+Statistics are enabled by default. They can be disabled at the compile time
+of the libvmemcache library if **VMEMCACHE_NO_STATS** is defined.
```
Comment (marcinslusarz) [inline README.md]: This should be a cmake option.

File: README.md [hunk 5]
```diff
@@ -30,3 +30,10 @@ $ cmake .. -DCMAKE_INSTALL_PREFIX=~/vmemcache-bin
 $ make
 $ make install
 ```
+
+# Statistics #
+
+Statistics are enabled by default. They can be disabled at the compile time
```
Comment (marcinslusarz) [inline README.md]: Do we really want to maintain both? Did we measure the impact of stats?

File: benchmarks/bench_simul.c (hunk 3 above)
Comment (ldorau) [reply benchmarks/bench_simul.c]: Done

File: README.md (hunk 4 above)
Comment (ldorau) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (ldorau) [reply README.md]: What do you mean by 'both'? Yes, it is a bit faster

Comment (kilobyte) [reply README.md]: Holy crap, on a 96-core, 2-socket machine: bench_simul type=index: with stats, 6.8 µs, without stats 1.3 µs!!!!!1!!1111!-e^(iπ)!!

Me wants.

Looks like we should look hard into decreasing the impact of gathering stats if we're to leave them on by default.

Comment (marcinslusarz) [reply README.md]: @ldorau by "both" I meant "both options".
@kilobyte how the numbers look like when you limit the benchmark to one numa node?

File: benchmarks/bench_simul.c (hunk 3 above)
Comment (marcinslusarz) [reply benchmarks/bench_simul.c]: Nope, you just renamed it and I think the new name is even worse (REMOVE_STATS sounds like vmemcache will actively remove stats now and that's not true).

STATS_ENABLED?

File: README.md (hunk 5 above)
Comment (ldorau) [reply README.md]: Both - yes

File: benchmarks/bench_simul.c (hunk 3 above)
Comment (ldorau) [reply benchmarks/bench_simul.c]: Done

File: README.md (hunk 5 above)
Comment (kilobyte) [reply README.md]: bench_simul n_threads=48 type=index

||both sockets|numactl -N 0|
|-|-|-|
|STATS|5.3|2.6|
|NO_STATS|0.8|0.52|

Comment (marcinslusarz) [reply README.md]: interesting, it seems we need per-thread stats...

**Comment by marcinslusarz** [comment]
We definitely need to do something about stats (disable them by default or lower their overhead), but this PR is good enough for now, so I'm merging this.

**Comment by ldorau** [comment]
> We definitely need to do something about stats (disable them by default or lower their overhead), but this PR is good enough for now, so I'm merging this.

@pbalcer Can we disable them by default?

**Pull request merged**